### PR TITLE
Feat/homepage restructuring

### LINKED ISSUE #100: Homepage Functionality
## User story
As a student, I want the homepage to be useful after I log in, so that I can 
quickly navigate to what I need without going through the dashboard every time.

## Context
The homepage currently shows only Login and Sign Up buttons when no user is 
logged in, and a single "Go to Dashboard" card when logged in. The WhatsApp 
discussion noted this has no functionality for a logged-in student. This story 
adds a quick-stats row for logged-in students (next consultation date, number 
of enrolled courses) and a shortcut card to find a consultation, so the homepage 
is a useful landing page rather than a redirect screen.

For lecturers, add a shortcut to set availability if none is set yet (onboarding 
nudge) or show their next upcoming consultation if they have one.

## Acceptance tests
- [ ] Logged-in student sees: next upcoming consultation date (or "None booked"), 
      number of enrolled courses, and a "Find a consultation" button
- [ ] Logged-in lecturer sees: next upcoming consultation (or "None booked") 
      and a shortcut to Set Availability
- [ ] Unauthenticated visitor sees existing Login / Sign Up layout unchanged
- [ ] Homepage loads in under 200ms (no N+1 queries — use a single aggregation 
      query per user type)

**Labels:** `user-story` `enhancement`  
**Story points:** 3  
**Sprint:** 4  
**Parent:** #3 (View consultation dashboards)

### PR DESCRIPTION
**Pull request by ThandiAlgorithm**
Related #100 

### Homepage Redesign

- Replaced the placeholder homepage with a fully designed landing page for KnockKnock.prof.
- Added a hero section with the door logo, brand name, tagline, and call-to-action buttons for student and lecturer login.
- Added a features section that explains the core value of the platform for both students and lecturers.
- Added a footer with the application name and university attribution.

### Homepage Controller

- Moved homepage rendering logic into `src/controllers/homepage-controller.js`.
- Added controller logic to show the homepage to unauthenticated users.
- Added role-based redirects so logged-in users are sent to the correct dashboard based on their session role.

### Brand Theme

- Added `public/css/theme.css` as a shared stylesheet used across the application.
- Defined reusable CSS variables for the green and orange brand colour palette, shadows, spacing, and borders.
- Added shared utility classes for the topbar, sidebar, stat cards, greetings, chips, and animations.
- Improved visual consistency across lecturer, student, and admin views by reducing duplicated CSS.

### Logo Assets

- Added `public/images/door-logo.svg`.
- Added `public/images/door-logo-transparent.svg`.
- Updated `public/images/logo.svg`.

### Testing

- Added unit tests for the homepage controller redirect logic.
- Added integration tests to confirm that the homepage renders correctly for unauthenticated users.
- Added integration tests to confirm that authenticated users are redirected to their correct dashboards.

## Why It Was Changed

The application previously opened directly on a login screen without giving users context about KnockKnock.prof. The new homepage introduces the brand, explains the purpose of the platform, and gives students and lecturers a clear entry point.

The homepage controller was added to keep routing logic separate from the route file, making the code cleaner and easier to maintain.

The shared theme stylesheet was introduced to keep colours, spacing, layout styles, and reusable UI elements consistent across the lecturer, student, and admin pages.

### CODE REVIEW

File: public/images/door-logo-transparent.svg
Comment (Aditya-Raghunandan) [inline public/images/door-logo-transparent.svg:1]: Such an artist

File: src/controllers/homepage-controller.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    return res.render('homepage', { user: null, stats: null })
+  }
+
+  const user = {
+    id: req.session.userId,
+    name: req.session.userName,
+    role: req.session.userRole,
+  }
+
+  const today = todayStr()
```
Comment (Aditya-Raghunandan) [inline src/controllers/homepage-controller.js:30]: This reimplements "today as YYYY-MM-DD" manually. SQLite's DATE('now') already returns exactly that format, and since db is already imported you could replace the whole function with a single db.prepare("SELECT DATE('now') AS d").get().d. The manual implementation also has a subtle timezone issue — new Date() uses the server's local timezone, so on a UTC-hosted Render instance the date rolls over at midnight UTC, not midnight SAST. Worth flagging even if it's not breaking anything today.

File: src/controllers/homepage-controller.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  return `${parseInt(d, 10)} ${MONTHS[parseInt(m, 10) - 1]}`
+}
+
+const todayStr = () => {
+  const now = new Date()
+  const pad = n => String(n).padStart(2, '0')
+  return `${now.getFullYear()}-${pad(now.getMonth() + 1)}-${pad(now.getDate())}`
+}
+
+const showHomepage = (req, res) => {
```
Comment (Aditya-Raghunandan) [inline src/controllers/homepage-controller.js:19]: The correlated subquery for nextDate runs a SELECT 1 FROM consultation_attendees inside an EXISTS that is itself inside a subquery inside SELECT. That's three levels of nesting for what is essentially a join. Consider flattening it: ,SELECT COUNT(*) AS courseCount,
       MIN(c.consultation_date || ' ' || c.consultation_time) AS nextSlot
FROM   consultations c
JOIN   consultation_attendees ca ON ca.const_id = c.const_id
WHERE  ca.student_number = ?
  AND  c.consultation_date >= ?
  AND  c.status IN ('Available', 'Booked', 'Ongoing')

File: src/controllers/homepage-controller.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      const row = db.prepare(`
+        SELECT
+          (SELECT consultation_date FROM consultations c
+           WHERE c.lecturer_id = ?
+             AND c.consultation_date >= ?
+             AND c.status IN ('Available', 'Booked', 'Ongoing')
+           ORDER BY c.consultation_date ASC, c.consultation_time ASC
+           LIMIT 1) AS nextDate,
+          (SELECT COUNT(*) FROM lecturer_availability WHERE staff_number = ?) AS availabilityCount
+      `).get(user.id, today, user.id)
```
Comment (Aditya-Raghunandan) [inline src/controllers/homepage-controller.js]: availabilityCount and nextDate are fetched in one get() call but they query completely independent tables with no join condition between them. Splitting them into two named queries (getNextConsultation and getAvailabilityCount) would make each one self-contained and individually testable, which matters given the rubric's emphasis on meaningful tests per story.

File: src/controllers/homepage-controller.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
+const db = require('../../database/db')
+
+const MONTHS = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
+
+const formatDate = (dateStr) => {
+  if (!dateStr) return null
+  const [, m, d] = dateStr.split('-')
+  return `${parseInt(d, 10)} ${MONTHS[parseInt(m, 10) - 1]}`
+}
```
Comment (Aditya-Raghunandan) [inline src/controllers/homepage-controller.js:13]: The destructuring const [, m, d] = dateStr.split('-') silently returns undefined if dateStr is not a properly formatted date string (e.g. a stray null that slipped past the early return, or a datetime string with a T). A short guard or a comment explaining the assumed format would prevent a future silent failure.

File: src/controllers/homepage-controller.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
+const db = require('../../database/db')
+
+const MONTHS = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
+
+const formatDate = (dateStr) => {
+  if (!dateStr) return null
+  const [, m, d] = dateStr.split('-')
+  return `${parseInt(d, 10)} ${MONTHS[parseInt(m, 10) - 1]}`
```
Comment (Aditya-Raghunandan) [inline src/controllers/homepage-controller.js]: Minor: MONTHS and formatDate are module-level constants but formatDate is also exported. If it's part of the public API (used by other controllers or tests), it should be documented as such. If it's only exported for testing, that's fine but worth a comment so the next person doesn't assume it's part of the controller's public contract.

**Review by Aditya-Raghunandan — COMMENTED** [review]
Good PR. The controller extraction is the right call , routing logic no longer bleeds into app.js, and the separation will make future changes to the homepage much cleaner. The shared theme.css with CSS variables is solid work; it will save real time as the app grows.
The main thing to revisit before merge is the todayStr() timezone issue — on Render (UTC), a student in Johannesburg (UTC+2) booking a consultation at 11pm SAST would see it disappear from "upcoming" two hours before their local midnight. That's a correctness bug, not just a style concern.
The nested SQL in the student query is also worth simplifying now rather than later, because when the acceptance tests grow to cover more edge cases (back-to-back bookings, same-day consultations) the current structure will be harder to reason about.
Tests look appropriate for the scope — unauthenticated visitor, logged-in student, logged-in lecturer with and without availability. The beforeAll/afterAll cleanup pattern matches the existing test conventions, which is good. One thing missing: a test that verifies the stats object is populated correctly for an authenticated student who has an upcoming consultation, not just that the dashboard link is present. That would push the integration test coverage from "renders without error" toward "renders correct data."

File: src/controllers/homepage-controller.js (hunk 1 above)
Comment (ThandiAlgorithm) [reply src/controllers/homepage-controller.js:30]: I will try fixing these issues. Thank you for flagging them

File: src/controllers/homepage-controller.js (hunk 2 above)
Comment (ThandiAlgorithm) [reply src/controllers/homepage-controller.js:19]: I will look into this

File: src/controllers/homepage-controller.js (hunk 3 above)
Comment (ThandiAlgorithm) [reply src/controllers/homepage-controller.js]: yes this makes a lot of sense.

File: src/controllers/homepage-controller.js (hunk 4 above)
Comment (ThandiAlgorithm) [reply src/controllers/homepage-controller.js:13]: I will look into this

File: src/controllers/homepage-controller.js (hunk 5 above)
Comment (ThandiAlgorithm) [reply src/controllers/homepage-controller.js]: okay
